sbt-bridge/src/xsbt/CompilerClassLoader.java

```diff
@@ -0,0 +1,122 @@
+package xsbt;
+
+import java.lang.reflect.Field;
+
+import java.net.URL;
+import java.net.URLClassLoader;
+
+import java.util.WeakHashMap;
+
+/**
+ * A classloader to run the compiler
+ * <p>
+ * A CompilerClassLoader is constructed from a list of `urls` that need to be on
+ * the classpath to run the compiler and the classloader used by sbt.
+ * <p>
+ * To understand why a custom classloader is needed for the compiler, let us
+ * describe some alternatives that wouldn't work.
+ * <ul>
+ * <li>`new URLClassLoader(urls)`:
+ *   The compiler contains sbt phases that callback to sbt using the `xsbti.*`
+ *   interfaces. If `urls` does not contain the sbt interfaces we'll get a
+ *   `ClassNotFoundException` in the compiler when we try to use them, if
+ *   `urls` does contain the interfaces we'll get a `ClassCastException` or a
+ *   `LinkageError` because if the same class is loaded by two different
+ *   classloaders, they are considered distinct by the JVM.
+ * <li>`new URLClassLoader(urls, sbtLoader)`:
+ *    Because of the JVM delegation model, this means that we will only load
+ *    a class from `urls` if it's not present in the parent `sbtLoader`, but
+ *    sbt uses its own version of the scala compiler and scala library which
+ *    is not the one we need to run the compiler.
+ * </ul>
+ * <p>
+ * Our solution is to implement a subclass of URLClassLoader with no parent, instead
+ * we override `loadClass` to load the `xsbti.*` interfaces from `sbtLoader`.
+ */
+public class CompilerClassLoader extends URLClassLoader {
+  private final ClassLoader sbtLoader;
+
+  public CompilerClassLoader(URL[] urls, ClassLoader sbtLoader) {
+    super(urls, null);
+    this.sbtLoader = sbtLoader;
+  }
+
+  @Override
+  public Class<?> loadClass(String className, boolean resolve) throws ClassNotFoundException {
+    if (className.startsWith("xsbti.")) {
+      // We can't use the loadClass overload with two arguments because it's
+      // protected, but we can do the same by hand (the classloader instance
+      // from which we call resolveClass does not matter).
+      Class<?> c = sbtLoader.loadClass(className);
+      if (resolve)
+        resolveClass(c);
+      return c;
+    } else {
+      return super.loadClass(className, resolve);
+    }
+  }
+
+  /**
+   * Cache the result of `fixBridgeLoader`.
+   * <p>
+   * Reusing ClassLoaders is important for warm performance since otherwise the
+   * JIT code cache for the compiler will be discarded between every call to
+   * the sbt `compile` task.
+   */
+  private static WeakHashMap<ClassLoader, ClassLoader> fixedLoaderCache = new WeakHashMap<>();
+
+  /**
+   * Fix the compiler bridge ClassLoader
+   * <p>
+   * Soundtrack: https://www.youtube.com/watch?v=imamcajBEJs
+   * <p>
```

Comment:
No `</p>` ?

Reply:
No, the Javadoc guide does not use the closing `</p>`s.